### PR TITLE
Upgrade to ECS 1.7.0

### PR DESCRIPTION
**Pull request by andrewstucki**
## What does this PR do?

This adds the upcoming ECS 1.7.0 yml file and updates all references to `ecs.version` -- also, it adds `ecs.version` data to the beats that were lacking it.

It's currently in draft because the final ECS 1.7.0 release isn't cut yet. But, once it is, if anything has changed, it should just be a matter of re-importing the `fields.ecs.yml` file and re-running `make update`.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: filebeat/docs/fields.asciidoc
Comment (andrewstucki) [inline filebeat/docs/fields.asciidoc:44308]: So, I'm not entirely sure how we want to document these -- do we need to change the doc generator to specify that sometimes this is a wildcard (if you're Elastic licensed), sometimes it's a keyword (if you're OSS)? Or should we add some sort of top-level documentation that says "all wildcards are keywords in OSS"?

@andrewkroh @webmat @ebeahan any thoughts?

Comment (webmat) [reply filebeat/docs/fields.asciidoc:44308]: Yeah we have an issue in ECS for 1.8 to document the "fallback" idea for OSS implementations, and what it means at a high level. We can probably cover the Beats and Logstash implementations there as well.

Perhaps Beats docs can link to that?

We can make sure this page is ready in advance of the Stack 7.11 release.

Comment (webmat) [reply filebeat/docs/fields.asciidoc:44308]: Just to answer your question a little more directly

> do we need to change the doc generator...

I don't think we need to. The "OSS" fallback is not an official variant of ECS, it's a best effort workaround for purely OSS implementations. People could even use that to adopt ECS on 6.x if they have complex setups with many versions. But we're not making tons of efforts to fully support this. For example basic types that don't really have a fallback will simply not be supported.

This may deserve a mention somewhere in the Beats documentation, but I'm not even sure if there's a section about OSS Beats.

Comment (andrewstucki) [reply filebeat/docs/fields.asciidoc:44308]: 👍 -- also, should we mention that we're using "experimental" features of ECS somewhere in our docs?

File: libbeat/_meta/fields.ecs.yml
Comment (webmat) [inline libbeat/_meta/fields.ecs.yml:2]: @ebeahan This is an import of the 1.7 + experimental changes.

I think it may be worth tweaking the header in our artifact to reflect that, WDYT?

Comment (andrewstucki) [reply libbeat/_meta/fields.ecs.yml:2]: I did do a slight no-no and modified it to say `1.7.0` v. `1.7.0-dev` -- but yeah, currently the experimental builds have nothing that indicates in the manifests that they're "experimental"

Comment (webmat) [reply libbeat/_meta/fields.ecs.yml:2]: Yeah I saw that 🧐 😉

File: filebeat/docs/fields.asciidoc
Comment (webmat) [reply filebeat/docs/fields.asciidoc]: Not sure. It's likely something that could flip flop from one release to another. For instance, "multi-user" stage 3 RFC is now merged, and will be part of mainline ECS in 1.8. Wildcard is approved, and it's a matter of time until it's merged as well.

So next release this notice may go away. And the following release, if we have another "stage 2" change we're pushing to adopt aggressively, the notice may need to come back... So I think it could add more confusion than anything.

It would be a fair warning to have though, so I wouldn't oppose if you want to add this.

File: libbeat/_meta/fields.ecs.yml
Comment (ebeahan) [reply libbeat/_meta/fields.ecs.yml:2]: I believe it's pulling from the `version` defined in the branch the artifacts are generated?

https://github.com/elastic/ecs/blob/1.7/experimental/generated/beats/fields.ecs.yml#L2

Comment (andrewstucki) [reply libbeat/_meta/fields.ecs.yml:2]: Yep: [from the generator](https://github.com/elastic/ecs/blob/6cbb5448edf921ed9b8b797bd8310003f109259a/scripts/generator.py#L89-L97). Wondering if it makes sense to add some of the `experimental` detection stuff from the [schema loader](https://github.com/elastic/ecs/blob/6cbb5448edf921ed9b8b797bd8310003f109259a/scripts/schema/loader.py#L60) into the generator script. That way you can have the context of whether this is an experimental build or not from the get-go -- then you'd be able to generate a string like `ECS version 1.7.0 (experimental)` in all of the generated files or something.

Either way, apart from people upgrading ECS in beats and not knowing we take from the experimental folder, this probably isn't that big of a deal.

File: filebeat/docs/fields.asciidoc
Comment (andrewkroh) [reply filebeat/docs/fields.asciidoc]: I do think this deserves a mention in the documentation so that users are aware that OSS beats will automatically be using `keyword` rather than `wildcard`. And that the same happens if you're using older, pre-wildcard, ES versions.

@dedemorton Do you have a recommendation on where to include such a notice?

Comment (dedemorton) [reply filebeat/docs/fields.asciidoc]: This is a tough call. It sounds like saying something in the field description (where it would be the most noticeable) is not a good idea because it would be repetitive (plus our docgen process limits us).

We can put a statement at the top-level page in the field reference for each Beat, but IMO it's likely to be overlooked by users. 

When are users most likely to care about the type difference? Will they be able to continue using the pre-built dashboards? Are there any upgrade implications if they decide to move to basic later on? Will there be a log message to let users know that keyword is being used instead of wildcard (and the implications/limitations)?

Comment (webmat) [reply filebeat/docs/fields.asciidoc]: @dedemorton `wildcard` is a drop-in replacement in terms of functionality. Everything will continue to work functionally the same (including dashboards). Only the performance characteristics are changing. So everything keeps working, some things faster, some things slower (check out the table in [the blog post for a breakdown](https://www.elastic.co/blog/find-strings-within-strings-faster-with-the-new-elasticsearch-wildcard-field)).

A user upgrading Basic Beats from 7.10 to 7.11 or from Beats OSS to Beats Basic 7.11+ will go through the same non-event. The `wildcard` and `keyword` types are compatible and mixed indices being queried together will work flawlessly. 

Users will only be able to fully take advantage of the performance improvements enabled by `wildcard` once the older data has aged out, though.

**Comment by webmat** [comment]
@andrewstucki ECS 1.7 is out. You can go grab the official v1.7.0 artifacts to update the PR :-)

File: libbeat/_meta/fields.ecs.yml
Comment (webmat) [inline libbeat/_meta/fields.ecs.yml:2]: Pristine file 👍  😉

Comment (andrewstucki) [reply libbeat/_meta/fields.ecs.yml:2]: 😅

File: filebeat/module/apache/access/config/access.yml
Comment (leehinman) [inline filebeat/module/apache/access/config/access.yml:11]: I'm wondering if we should wait to bump up the ecs version in each of the beats/modules until we have evaluated them.  I've taken this value to mean that we have looked at the module and to the best of our ability we are mapping values to the available ECS fields in this version.  So if Apache says it is ECS 1.7.0, then I might expect it to have `url.subdomain` (new ECS 1.7 field) mapped, where if it was reporting ECS 1.6.0 I would know that it wouldn't have that field.  
Thoughts?

Comment (andrewstucki) [reply filebeat/module/apache/access/config/access.yml:11]: So, generally I'd agree with that, in this case though one of the large changes for 1.7.0 experimental is that we get the remapping of a bunch of fields from `keyword` to `wildcard` -- in this case the fact that that's such a major mapping change makes me think we should just update all to 1.7.0 initially and just track the additional field enhancements in an issue. At that point we're using this to signal "yeah, we're shipping 1.7.0 schema" even though we don't have all final fields filled in yet before the release. The alternative, saying "we're shipping 1.6.0, but these fields are actually 1.7.0 experimental wildcards" seems less ideal to me.

Comment (leehinman) [reply filebeat/module/apache/access/config/access.yml:11]: darn wildcards 😄 

that makes sense.

**Review by leehinman — APPROVED** [review]

File: filebeat/module/apache/access/config/access.yml
Comment (webmat) [reply filebeat/module/apache/access/config/access.yml:11]: darn grey areas 😂 

One thing to consider is that the `ecs.version` can either be a statement about:

* which fields are populated in the event
* the mapping itself. Like whether it uses `keyword` or `wildcard`

I've always taken it to mean the former. Although I'm thinking users may want to start doing queries that depend on the presence of `wildcard`, and want to exclude older indices. Using `ecs.version` for that would probably be the easiest way.

On the "shape of the event" aspect, we should consider that ECS 1.7 introduced the following fields:

* Mainline ECS: `http.[request|response].mime_type` and `threat.technique.subtechnique`
* Experimental: `user.target.*`, `user.effective.*` and `user.changes.*`


In general, I would favor the conservative approach of only updating that number when the source has been adjusted to populate new relevant fields (also ok to upgrade if the source is unaffected).

But in this case, I would err on the side of updating the version to 1.7.0 aggressively, so that users can use `ecs.version` to guard potentially expensive queries from hitting older `keyword`-based indices.

Comment (andrewstucki) [reply filebeat/module/apache/access/config/access.yml:11]: @webmat I like that general guidance. I'll see about documenting that somewhere in the repo in developer docs in a subsequent PR. If this weren't quite so large of a change, definitely agree that it would make sense just to update as we added the fields.

**Comment by webmat** [comment]
@andrewstucki We found a problem in the Beats YML file today, relating to the APM and ECS logger "tracing" fields. In Beats they're added under a "tracing" section, which makes them incorrectly appear as "tracing.trace.id", "tracing.transaction.id" and so on.

We have a PR to address this issue https://github.com/elastic/ecs/issues/1163 already. I think it's worth waiting for it to be merged and pull this bug fix in, before merging this PR.

**Comment by webmat** [comment]
@andrewstucki Alright, the fix has been merged in ECS branch "1.7". You can go grab it.

Note that [the tracing fields](https://github.com/elastic/ecs/blob/1.7/experimental/generated/beats/fields.ecs.yml#L5170-L5197) are now at the same level in the YAML document as field groups usually are. The field groups generally don't have a dot in their name, only fields defined within them do... So I'm not 100% sure if this will work as is. Please let us know if this causes an issue, we can adjust it further. Sorry I only noticed/ thought about that after it was backported.

**Review by webmat — APPROVED** [review]
Latest import and trace.id correction LGTM 👍

**Pull request merged**
